test/e2utils/kpm_utils.go

```diff
@@ -22,76 +24,171 @@ const (
 	defaultGranularity  = uint32(500)
 )
 
+// Sub represents the common fields for subscriptions
 type Sub struct {
-	Name      string
-	NodeID    topo.ID
-	Node      sdkclient.Node
-	SdkClient sdkclient.Client
-	Ch        chan e2api.Indication
+	Name                string
+	NodeID              topo.ID
+	Node                sdkclient.Node
+	SdkClient           sdkclient.Client
+	Ch                  chan e2api.Indication
+	Timeout             time.Duration
+	Actions             []e2api.Action
+	EventTriggerBytes   []byte
+	ActionID            int32
+	ActionType          e2api.ActionType
```

Comment:
We can have multiple actions per subscription. I see you already defined a list of actions but Action Type and Action ID are defined separately. Action https://github.com/onosproject/onos-api/blob/master/proto/onos/e2t/e2/v1beta1/subscription.proto#L57 already has action ID and type

Reply:
ActionID and ActionType here are used when the caller does not specify any actions and a default action is created (most of the tests do this)

Reply:
I guess we need to change that then. At least one Action should be specified part of the request otherwise the request is not valid. There is no such thing as default action. For the tests, mostly the action is REPORT.

Reply:
That is true the Action can be eliminated part of the protobuf request but from E2 node persecutive the request should have one.

Reply:
A default action is created by the Subscribe() function if none is specified.